### ISSUE TITLE
schema specification shouldn't rely on object key order

### ISSUE THREAD

**Issue by blipboard** [jed/dynamo#9]
The current API pays relies key order in the schema parameter passed to new db.Table().

However, this can fail on an ECMA-262 5th Ed. complaint interpreter.  Perhaps it should be changed to a list:

``` javascript
recipeTable = new db.Table({
   name:  "recipes",
   schema: [ {userId: Number}, {date: String} ] // key order insensitive
   throughput: {read: 10, write: 10}
})
```

**Comment by jed** [comment]
this library is node-only, which means all we need to do is confirm to v8's handling, which iterates in insertion order EXCEPT when numbers are used as keys. that's really the only case where i could see this fail, and it is pretty rare. providing a dead-easy API with a fallback to the original raw API for when this isn't good enough is a pretty good compromise, i think, but am willing to entertain pull requests that add the ability to use arrays without affecting the current API.

**Comment by blipboard** [comment]
I asked about this on [StackOverflow](http://stackoverflow.com/questions/9179680/is-it-acceptable-style-for-node-js-libraries-to-rely-on-object-key-order), and opinion seems to be against doing this for node libraries.  I think the problem arises when (for performance reasons), V8 decides to change its implementation.  Libraries which build against the current API will break.  

Another problem is that relying on an undocumented behavior in an API specification may cause confusion for some programmers.  Ordered keys in hashtables are an unusual behavior, and a reader of dynamo.js schemas who isn't familiar with these semantics may not get what is happening.

(What file is this code in, btw?  Maybe I'll take a crack at it.)

**Comment by jed** [comment]
wow, some opinionated people in that thread.

i'd like to add support for more explicit key order, but there are several options:
- accepting an array
- accepting a hash with `hash` and `range` keys that have `Object` values
- requiring a schema to be specified when a table is created

i think the first is the most straightforward.

there are two places that would need attention: [Key.js](https://github.com/jed/dynamo/blob/master/lib/Key.js#L3-11) and [KeySchema.js](https://github.com/jed/dynamo/blob/master/lib/KeySchema.js#L4-10). i'm in favor or allowing arrays using `isArray` to check. would you like to take a crack at it?

**Comment by blipboard** [comment]
Excellent.  I will try to take a look, but can't promise anything quickly.   (Or if someone can get to it before me, please do)

**Comment by mlogan** [comment]
Hi everyone,

I took a crack at this: See https://github.com/jed/dynamo/pull/35

Comments welcome.

Thanks,
Mark

**Comment by jed** [comment]
fixed by @mlogan in #35.
